### ISSUE TITLE
Decimal numbers in BPM

### ISSUE THREAD

**Issue by dozzzzer** [audacity/audacity#4457]
### Your idea

My idea is to accept decimal numbers with at least 1 digit in the decimal part as **Tempo** values. Which, in turn, will require the Tempo text box to be made slightly wider to accommodate the decimal part too.

### Problem to be solved

In music production, often times one may need to align some of the instruments/audio tracks to a loop/phrase with the **Tempo** value twice as much. For instance, a loop of 177.0 BPM needs to be in sync with some other instruments that are desired to be half the tempo of it, i.e. _88.5 BPM_. Which makes it not feasible given it only accepts whole numbers.

**Comment by Sreejan100** [comment]
@dozzzzer  hey I would want to solve this issue . We have to just modify the input box to accept decimal values right?

**Comment by dozzzzer** [comment]
> We have to just modify the input box to accept decimal values right?

I'm afraid I'm not the right person to ask this :)

**Comment by crsib** [comment]
Should be fixed by #4414

**Comment by DavidBailes** [comment]
I've tried putting a decimal point in the tempo, but it's ignored. Is this just me?

**Comment by crsib** [comment]
I think this is a bug in wxWidgets. I can enter decimal points from the numpad without any issues.

**Comment by DavidBailes** [comment]
> I think this is a bug in wxWidgets. I can enter decimal points from the numpad without any issues.

I don't think so. It's a keyboard event issue in Audacity. If in preferences > keyboard, you remove the shortcut "." from cursor short jump right, you can then type "." in the tempo control.

**Comment by crsib** [comment]
Well, that's worse...

**Comment by DavidBailes** [comment]
> Well, that's worse...

It's a little while since I had a look at the key event handling code in detail, and there may be better ways of doing things, but...

The way that NumericTextCtrl makes sure it gets the keystrokes it wants, before they are sent to see if they are assigned to commands, is to register that it wants first go for a set of keystrokes. See:
void NumericTextCtrl::OnFocus(wxFocusEvent &event).
Some keystrokes are automatically kept away from their possible use as commands. See:
bool CommandManager::FilterKeyEvent(AudacityProject *project, const wxKeyEvent & evt, bool permit)

**Comment by crsib** [comment]
Hmm, I'll take a look, thank you!

**Comment by crsib** [comment]
Turns out to be quite complex: 

`KeyboardCapture` filters by `ON_KEY_DOWN`, but wxTextCtrl does not consume `.` in `ON_KEY_DOWN`, using `ON_CHAR` instead.

**Comment by DavidBailes** [comment]
> Turns out to be quite complex:
> 
> `KeyboardCapture` filters by `ON_KEY_DOWN`, but wxTextCtrl does not consume `.` in `ON_KEY_DOWN`, using `ON_CHAR` instead.

Yes, ideally the text box should always get the "." or "," characters based on the Unicode values rather than virtual key codes, so that it works on all keyboard layouts. You may have already come up with a good fix for this, but just wanted to share the following observations.

There is a similar, but not identical issue in some of the built in Audacity effects when they are used as realtime effects. For example, in the Bass and Treble effect, you should be able to enter a decimal separator in the text boxes and you can. In FilterEvent(wxEvent& event) in KeyboardCapture.cpp, the keyboard event is given back to WX for normal processing, since the commands which use the . and , shortcuts are not enabled because the track panel is not the focus.

However the Tempo text box is in the main window, and so (only on Windows I think), before wxWidgets gives the keyboard event to to text box, it checks for accelerators in the menus. I finds one, and so the text box doesn't get the keyboard event.
If in CommandManager::FormatLabelWithDisabledAccel(const CommandListEntry *entry) const, you could add the lines:
if( key.StartsWith(".")) break;
if( key.StartsWith(",")) break;
Then the Tempo text box would work in a similar way to the text boxes in the realtime effect dialogs.
This isn't an ideal fix, but will work in most keyboard layouts - see:
http://kbdlayout.info/features/virtualkeys/VK_OEM_COMMA
http://kbdlayout.info/features/virtualkeys/VK_OEM_PERIOD
( I guess https://github.com/audacity/audacity/issues/4390 is also relevant).

**Comment by crsib** [comment]
Ah, yes, I guess accel keys handling by wx is what I am missing

**Comment by DavidBailes** [comment]
> https://forums.wxwidgets.org/viewtopic.php?t=20221

but the solution described there is using virtual keycodes, rather than the unicode value of characters. The key combinations for "." and "," won't have the same virtual keycodes on all the regional keyboard layouts (though they will on most).

**Comment by crsib** [comment]
I am only trying to have a workaround for the specific issue, not to fix a global problem with shortcuts. I don't think that the shortcuts problem will be addressed in Audacity 3.

**Comment by DavidBailes** [comment]
OK.

**Comment by crsib** [comment]
Seems the solution you proposed is the best, thank you a lot!

**Comment by DavidBailes** [comment]
> Seems the solution you proposed is the best, thank you a lot!

Alas, that solution only works when the Extra menu is shown in the menubar. No idea why!

**Comment by dozzzzer** [comment]
Moving back to In Progress: decimal numbers still cannot be entered in the Tempo text box.

**Comment by DavidBailes** [comment]
@crsib . I've had another look at this morning, and just letting you know what I've found so far, in case you're also looking at this. I'm just using the unmodified beta.

I rechecked entering "." in the text boxes in Audacity realtime effects, such as base and treble. You can only enter "." if the extra menu is shown.
Checking in FilterEvent(wxEvent& event) in KeyboardCapture.cpp to see what was happening in this case, then in the lines:
if ( KeyboardCapture::PostFilter::Call( key ) )
            return Event_Processed;
Then for a keydown event, when Extra menus are shown, return Event_Processed is correctly not executed, but when Extra menus are not shown, the line is (I think) incorrectly executed, and so the event isn't passed to wx.

**Comment by crsib** [comment]
Thank you a lot David!

@dozzzzer FYI ^^

**Comment by DavidBailes** [comment]
continuing the story...
So in bool CommandManager::HandleCommandEntry()
If the Extra menu is shown, then entry->enabled is false, and so the function returns false.
If the Extra menu is not shown, entry-enabled is true (which I think is incorrect), allowed is false, and the function returns true.

**Comment by DavidBailes** [comment]
... if fixing the updating of the enabling of the commands on the Extra menu when it is hidden is not straight forward, a simpler alternative fix/hack would be to add some lines to CommandManager::FilterKeyEvent():
After the end of the existing cases of the switch statement:
switch( evt.GetKeyCode() ){

add some extra cases, so it would look like:
```
         case '7':
         case '8':
         case '9':
            return false;
         case '.':
         case ',':
            if (!evt.HasAnyModifiers())
               return false;
         }

```

**Comment by dozzzzer** [comment]
Tested
